### PR TITLE
Support array on parquet

### PR DESCRIPTION
**Pull request by DorianZheng**
- Create Hive table
```
create table person(name string, work_locations array<array<string>>)
ROW FORMAT DELIMITED
FIELDS TERMINATED BY '\t'
COLLECTION ITEMS TERMINATED BY ','
STORED AS PARQUET;
```

- Cteate StarRocks external table
```
REATE EXTERNAL TABLE `person` ( `name` varchar(200), `work_locations` array<array<varchar(200)>>) ENGINE=HIVE PROPERTIES (   "resource" = "hive",   "database" = "myhive",   "table" = "person" );
```

- Insert records into Hive table
```
insert into person select 'zzq', array(array('wuhan', 'hongkong'), array('beijing', 'tianjin'), array('guangzhou'), array('shanghai', 'hanzhou', 'shenzhen'));
insert into person  select 'zzq', array(array(), array('beijing', NULL), NULL, array('shanghai', '', 'shenzhen'));
```

- Query from StarRocks external table
```
select * from person;
```

- Results
```
mysql> select * from person;
+------+----------------------------------------------------------------------------------------------+
| name | work_locations                                                                               |
+------+----------------------------------------------------------------------------------------------+
| zzq  | [["wuhan","hongkong"],["beijing","tianjin"],["guangzhou"],["shanghai","hanzhou","shenzhen"]] |
| zzq  | [[],["beijing",null],null,["shanghai","","shenzhen"]]                                        |
+------+----------------------------------------------------------------------------------------------+
2 rows in set (0.02 sec)
```

### CODE REVIEW

**Comment by DorianZheng** [comment]
I will change `IcebergTable::validate` later

**Review by dirtysalt — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/catalog/HiveTable.java [hunk 1]
```diff
@@ -414,47 +413,54 @@ private void validate(Map<String, String> properties) throws DdlException {
         }
     }
 
-    private Set<PrimitiveType> getValidColumnType(String hiveType) {
+    private boolean validateColumnType(String hiveType, Type type) {
```
Comment (imay) [inline fe/fe-core/src/main/java/com/starrocks/catalog/HiveTable.java:416]: It seems that it's better to create HiveType class.

File: be/src/column/column.h [hunk 2]
```diff
@@ -332,6 +332,8 @@ class Column {
 
     virtual Status accept_mutable(ColumnVisitorMutable* visitor) = 0;
 
+    virtual Column* mutable_child_column() { return nullptr; }
```
Comment (imay) [inline be/src/column/column.h:335]: Can you give this function some explanation.
But I think you can create a more general function interface.
If we support Map or Struct Column type later, this function will be a little confusing.

File: fe/fe-core/src/main/java/com/starrocks/catalog/HiveTable.java (hunk 1 above)
Comment (DorianZheng) [reply fe/fe-core/src/main/java/com/starrocks/catalog/HiveTable.java:416]: Now we just do some type match validation, but I agree to create HiveType class if we have further use case on it.

File: be/src/exec/parquet/column_reader.cpp [hunk 3]
```diff
@@ -634,9 +645,23 @@ class ListColumnReader : public ColumnReader {
         offsets[0] = 0;
         def_rep_to_offset(_field->level_info, def_levels, rep_levels, num_levels, &offsets[0], &is_nulls[0],
                           &num_offsets);
-        return Status::OK();
+
+        for (int i = 1; i <= num_offsets; i++) {
+            array_column->offsets_column()->append(offsets[i]);
```
Comment (imay) [inline be/src/exec/parquet/column_reader.cpp:650]: This is a cost operation cause there are too many function calls.
I think better to get offsets_column and assigned by offsets?

File: be/src/exec/parquet/column_reader.cpp [hunk 4]
```diff
@@ -634,9 +645,23 @@ class ListColumnReader : public ColumnReader {
...
+        }
+        if (_field->is_nullable) {
+            DCHECK(dst->is_nullable());
+            DCHECK_NOTNULL(nullable_column);
+            for (int i = 1; i <= num_offsets; i++) {
```
Comment (imay) [inline be/src/exec/parquet/column_reader.cpp:655]:
```suggestion
            for (int i = 0; i < num_offsets; i++) {
```

**Comment by DorianZheng** [comment]
@imay PTAL

**Comment by DorianZheng** [comment]
run starrocks_fe_unittest

**Review by imay — DISMISSED** [review]
LGTM

**Comment by DorianZheng** [comment]
run starrocks_be_unittest

File: fe/fe-core/src/main/java/com/starrocks/catalog/IcebergTable.java [hunk 5]
```diff
@@ -173,41 +170,55 @@ private void validate(Map<String, String> properties) throws DdlException {
...
-                return Sets.newHashSet(PrimitiveType.DECIMALV2, PrimitiveType.DECIMAL32,
-                        PrimitiveType.DECIMAL64, PrimitiveType.DECIMAL128);
+                return primitiveType == PrimitiveType.DECIMALV2 ||
+                        primitiveType == PrimitiveType.DECIMAL32 ||
+                        primitiveType == PrimitiveType.DECIMAL64 ||
+                        primitiveType == PrimitiveType.DECIMAL128;
             case FIXED:
             case BINARY:
             case STRUCT:
             case LIST:
```
Comment (caneGuy) [inline fe/fe-core/src/main/java/com/starrocks/catalog/IcebergTable.java:218]: this case conflict with icebergType.isListType() checker previous?

**Comment by caneGuy** [comment]
should we test with iceberg table?

File: fe/fe-core/src/main/java/com/starrocks/catalog/HiveTable.java [hunk 6]
```diff
@@ -414,47 +413,54 @@ private void validate(Map<String, String> properties) throws DdlException {
...
             case "DOUBLE PRECISION":
-                return Sets.newHashSet(PrimitiveType.DOUBLE);
+                return type.getPrimitiveType() == PrimitiveType.DOUBLE;
             case "DECIMAL":
             case "NUMERIC":
-                return Sets.newHashSet(PrimitiveType.DECIMALV2, PrimitiveType.DECIMAL32, PrimitiveType.DECIMAL64,
-                        PrimitiveType.DECIMAL128);
+                return type.getPrimitiveType() == PrimitiveType.DECIMALV2 || type.getPrimitiveType() ==
+                        PrimitiveType.DECIMAL32 || type.getPrimitiveType() == PrimitiveType.DECIMAL64 ||
+                        type.getPrimitiveType() == PrimitiveType.DECIMAL128;
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/catalog/HiveTable.java:442]: Better to wrap the line after `||`.

File: fe/fe-core/src/main/java/com/starrocks/catalog/IcebergTable.java (hunk 5 above)
Comment (DorianZheng) [reply fe/fe-core/src/main/java/com/starrocks/catalog/IcebergTable.java:218]: Not really, this case would never reach by define

**Comment by DorianZheng** [comment]
> should we test with iceberg table?

I will test against iceberg table with creating external table case, the other case expect to be the same with hive table.

**Review by dirtysalt — APPROVED** [review]

**Review by wyb — APPROVED** [review]

**Pull request merged**